### PR TITLE
Adding phase to simplify enum and record matching statements

### PR DESCRIPTION
**Pull request by cmester0**
Implementation of #168 , #180 closed and moved here instead.

### CODE REVIEW

File: engine/lib/ast.ml [hunk 1]
```diff
@@ -271,13 +271,16 @@ functor
...
     and pat' =
       | PWild
       | PAscription of { typ : ty; typ_span : span; pat : pat }
       | PConstruct of {
           name : global_ident;
           args : field_pat list;
-          is_record : bool; (* are fields named? *)
+          is_record : F.project_instead_of_match option;
+          (* are fields named? *)
+          (* F.project_instead_of_match *)
```
Comment (W95Psp) [inline engine/lib/ast.ml:283]: cleanup?

File: engine/lib/phases/phase_project_instead_of_match.ml [hunk 2]
```diff
@@ -0,0 +1,227 @@
...
+              captures = List.map ~f:dexpr captures;
+            }
+
+    and darm' (_span : span) (a : A.arm') : B.arm' =
+      let simple_pat, remaining_pats = project_pat a.arm_pat in
+      {
+        arm_pat = simple_pat;
+        body = lets_of_pat_bindings remaining_pats (dexpr a.body);
+      }
+      [@@inline_ands bindings_of dexpr]
```
Comment (W95Psp) [inline engine/lib/phases/phase_project_instead_of_match.ml:220]: Mmh, I'm a bit worried about that one.
If I understand correctly, here something like:
```rust
match x {
  Some(Foo {a: None, b}) => b,
  Some(Foo {a: Some(a), b}) => a + b,
  None => 0,
}
```
would be rewritten into:
```rust
match x {
  Some(some_name) =>
    match some_name {
      Foo {a: some_name, b} => match some_name { None => b }
    },
  Some(some_name) => 
     match some_name {
       Foo {a: some_name, b} => match some_name { Some(a) => a + b},
     },
  None => 0,
}
```
am I right?

But then, the first sub-sub-match is incomplete and the second arm won't be reached.

File: engine/lib/phases/phase_project_instead_of_match.ml [hunk 3]
```diff
@@ -0,0 +1,227 @@
...
+        B.pat' * (B.pat * B.expr) list =
+      match pat with
+      | PWild -> (PWild, [])
+      | PAscription { typ; typ_span; pat } ->
+          let simple_pat, remaining_pats = project_pat pat in
+          ( PAscription { typ = dty span typ; pat = simple_pat; typ_span },
+            remaining_pats )
+      | PConstruct { name; args; is_record = Some _; is_struct } ->
+          let update_args = List.map ~f:(project_field_pat span) args in
+          let new_id = UA.fresh_local_ident_in_expr [] "some_name" in
```
Comment (W95Psp) [inline engine/lib/phases/phase_project_instead_of_match.ml:65]: `fresh_local_ident_in_expr` gives you fresh things only relative to a given list of expressions.
So here, you have no guarantees.
When a pattern has multiple sub-patterns, this is an issue, i.e. `match x {Foo {a: PAT_A, b: PAT_B} | ...}`. Currently `project_pat'` would generate two `some_name` bindings.

I guess one solution would be `project_pat'` to take an index or something.

File: engine/lib/phases/phase_project_instead_of_match.ml [hunk 4]
```diff
@@ -0,0 +1,227 @@
...
+                                subpat = None;
+                              };
+                          span;
+                          typ = TApp { ident = name; args = [] };
+                        };
+                    };
+                  ];
+                is_record = (None : FB.project_instead_of_match option);
+                is_struct;
+              },
```
Comment (W95Psp) [inline engine/lib/phases/phase_project_instead_of_match.ml:92]: Should that actually be a `PConstruct`? Here we rewrite `Foo {a: ..., b: ..., c: ...}` into `Foo {Foo: some_name}`, that's wrong right?

Comment (W95Psp) [reply engine/lib/phases/phase_project_instead_of_match.ml:92]: I feel like this should be rather be a `PBinding`

**Review by W95Psp — CHANGES_REQUESTED** [review]
Thanks!
I have some comments, I think some edge cases won't be handled correctly.

Can you add a test for this phase also?

File: engine/lib/phases/phase_project_instead_of_match.ml (hunk 3 above)
Comment (cmester0) [reply engine/lib/phases/phase_project_instead_of_match.ml:65]: Oh yes, I forgot to add the expressions here. I should probably make sure the variable is fresh compared to the expression it is being used in.

File: engine/lib/phases/phase_project_instead_of_match.ml (hunk 2 above)
Comment (cmester0) [reply engine/lib/phases/phase_project_instead_of_match.ml:220]: Currently a similar problem exists for
```rust
#![allow(dead_code)]
pub struct Foo {
    pub a : Option<usize>,
    pub b : usize,
}

pub fn bar(x : Option<Foo>) -> usize {
    match x {
        Some(f) if match f { Foo {a: None, b} => true, _ => false } => {
            let Foo {a: None, b} = f;
            b
        }
        Some(f) if match f { Foo {a: Some(a), b} => true, _ => false } => {
            let Foo {a: Some(a), b} = f;
            a + b
        }
        None => 0,
    }
}
```
which is the correct "dumb" way to do the translation. However I should fix this, such that multiple cases for the same constructor are respected (which is currently not the case).

Comment (cmester0) [reply engine/lib/phases/phase_project_instead_of_match.ml:220]: This is indeed an issue, I should have some check about which case I end up in before binding the values.

Comment (cmester0) [reply engine/lib/phases/phase_project_instead_of_match.ml:220]: I would expect the translation to look something like:
```rust
match x {
  Some(f) if if let (None, b) = (f.a, f.b) { true } else { false } => {
    let None = f.a;
    let b = f.b;
    b
  }
  None => 0,
  Some(f) if if let (Some(a), b) = (f.a, f.b) { true } else { false } => {
    let Some(a) = f.a;
    let b = f.b;
    a + b
  },
}
```
However we currently do not translate the conditions on match cases?

**Comment by cmester0** [comment]
I am having a hard time actually fixing the issues, without the use of conditional match statements, which are currently not supported. This should probably fall under the same or a very similar phase to removing conditional matching as most of the issues are the same. For now I use very few or none of the features, however disables any matching on structs for the ssprove backend.

**Comment by cmester0** [comment]
This was more difficult than expected, and not too relevant yet, thus we will close this for now.
